interop/lib/src/client.dart

```diff
@@ -0,0 +1,1027 @@
+  /// UNIMPLEMENTED status code.
+  ///
+  /// Procedure:
+  /// * Client calls `grpc.testing.TestService/UnimplementedCall` with an empty
+  ///   request (defined as `grpc.testing.Empty`):
+  ///     {
+  ///     }
+  ///
+  /// Client asserts:
+  /// * received status code is 12 (UNIMPLEMENTED)
+  Future<Null> unimplementedMethod() async {
+    try {
+      await client.unimplementedCall(new Empty());
+      throw 'Did not throw.';
+    } on GrpcError catch (e) {
+      if (e.code != StatusCode.unimplemented) {
+        throw 'Unexpected status code ${e.code} - ${e.message}.';
+      }
+    }
+  }
+
+  /// This test verifies calling an unimplemented server returns the
+  /// UNIMPLEMENTED status code.
+  ///
+  /// Procedure:
+  /// * Client calls `grpc.testing.UnimplementedService/UnimplementedCall` with
+  ///   an empty request (defined as `grpc.testing.Empty`)
+  ///
+  /// Client asserts:
+  /// * received status code is 12 (UNIMPLEMENTED)
+  Future<Null> unimplementedService() async {
+    try {
+      await unimplementedServiceClient.unimplementedCall(new Empty());
+      throw 'Did not throw.';
+    } on GrpcError catch (e) {
+      if (e.code != StatusCode.unimplemented) {
+        throw 'Unexpected status code ${e.code} - ${e.message}.';
+      }
+    }
+  }
+
+  /// This test verifies that a request can be cancelled after metadata has been
+  /// sent but before payloads are sent.
+  ///
+  /// Procedure:
+  /// 1. Client starts StreamingInputCall
+  /// 2. Client immediately cancels request
+  ///
+  /// Client asserts:
+  /// * Call completed with status CANCELLED
+  Future<Null> cancelAfterBegin() async {
+    final requests = new StreamController<StreamingInputCallRequest>();
+    final call = client.streamingInputCall(requests.stream);
+    scheduleMicrotask(call.cancel);
+    try {
+      await call;
+      throw 'Expected exception.';
+    } on GrpcError catch (e) {
+      if (e.code != StatusCode.cancelled) {
+        throw 'Unexpected status code ${e.code} - ${e.message}';
+      }
+    }
+    requests.close();
+  }
+
+  /// This test verifies that a request can be cancelled after receiving a
+  /// message from the server.
+  ///
+  /// Procedure:
+  /// 1. Client starts FullDuplexCall with
+  ///     {
+  ///       response_parameters: {
+  ///         size: 31415
+  ///       }
+  ///       payload: {
+  ///         body: 27182 bytes of zeros
+  ///       }
+  ///     }
+  ///
+  /// 2. After receiving a response, client cancels request
+  ///
+  /// Client asserts:
+  /// * Call completed with status CANCELLED
+  Future<Null> cancelAfterFirstResponse() async {
+    final requests = new StreamController<StreamingOutputCallRequest>();
+    final call = client.fullDuplexCall(requests.stream);
+    final completer = new Completer();
+
+    var receivedResponse = false;
+    call.listen((response) {
+      if (receivedResponse) throw 'Received too many responses.';
```

Comment:
Who receives the exception thrown here? Should this not complete `completer` with an error instead? Same on line 969.

Reply:
No-one can receive this exception, so yeah, it's probably better to complete with an error. (End result is the same, since no-one catches the error thrown by `await completer.future`, but at least someone *could*.)